// railway-api-wrapper/index.js
var request = require("request");

exports.apiKey = null;

var scanResponseCode = function(responseCode, callback){
    if (responseCode === 200) callback(null);
    else if (responseCode === 401) callback("Error: Authentication error. Either your API key was invalid or you passed a wrong hash.");
    else callback("Error: Quota exhausted for day");
} 

exports.setApiKey = function (key){
            apiKey = key;
}

exports.checkPNR = function (pnrNo,callback){
    request(
        `http://api.railwayapi.com/v2/pnr-status/pnr/${pnrNo}/apikey/${apiKey}/`,
        (err, res) => {
            if (err){
                throw err;
            }
            else{
                var response = JSON.parse(res.body);
                scanResponseCode(
                    response.response_code,
                    (resErr) => {
                        callback(resErr, response);
                    }
                );
            }
        }    
    );
}

exports.liveTrainStatus = function(trainNo,date,callback){
    request(
        `http://api.railwayapi.com/v2/live/train/${trainNo}/date/${date}/apikey/${apiKey}/`,
        (err, res) => {
            if (err){
                throw err;
            }
            else{
                var response = JSON.parse(res.body);
                scanResponseCode(
                    response.response_code,
                    (resErr) => {
                        callback(resErr, response);
                    }
                );
            }
        }    
    );
}

exports.stationName = function(srcStationCode,callback){
    request(
        `http://api.railwayapi.com/v2/code-to-name/code/${srcStationCode}/apikey/${apiKey}/`,
        (err, res) => {
            if (err){
                throw err;
            }
            else{
                var response = JSON.parse(res.body);
                scanResponseCode(
                    response.response_code,
                    (resErr) => {
                        callback(resErr, response);
                    }
                );
            }
        }    
    );
}

exports.stationCode = function(srcStationName,callback){
    request(
        `http://api.railwayapi.com/v2/name-to-code/station/${srcStationName}/apikey/${apiKey}/`,
        (err, res) => {
            if (err){
                throw err;
            }
            else{
                var response = JSON.parse(res.body);
                scanResponseCode(
                    response.response_code,
                    (resErr) => {
                        callback(resErr, response);
                    }
                );
            }
        }    
    );
}

exports.trainRoute = function(trainNo,callback){
    request(
        `http://api.railwayapi.com/v2/route/train/${trainNo}/apikey/${apiKey}/`,
        (err, res) => {
            if (err){
                throw err;
            }
            else{
                var response = JSON.parse(res.body);
                scanResponseCode(
                    response.response_code,
                    (resErr) => {
                        callback(resErr, response);
                    }
                );
            }
        }    
    );
}

exports.trainBetweenStations = function(srcStationCode,destStationCode,date,callback){
    request(
        `http://api.railwayapi.com/v2/between/source/${srcStationCode}/dest/${destStationCode}/date/${date}/apikey/${apiKey}/`,
        (err, res) => {
            if (err){
                throw err;
            }
            else{
                var response = JSON.parse(res.body);
                scanResponseCode(
                    response.response_code,
                    (resErr) => {
                        callback(resErr, response);
                    }
                );
            }
        }    
    );
}

exports.nameOrNumber = function(trainNameOrNum,callback){
    request(
        `http://api.railwayapi.com/v2/name-number/train/${trainNameOrNum}/apikey/${apiKey}/`,
        (err, res) => {
            if (err){
                throw err;
            }
            else{
                var response = JSON.parse(res.body);
                scanResponseCode(
                    response.response_code,
                    (resErr) => {
                        callback(resErr, response);
                    }
                );
            }
        }    
    );
}

exports.seatAvailability = function(trainNo,srcStationCode,destStationCode,date,classCode,quotaCode,callback){
    request(
        `http://api.railwayapi.com/v2/check-seat/train/${trainNo}/source/${srcStationCode}/dest/${destStationCode}/date/${date}/class/${classCode}/quota/${quotaCode}/apikey/${apiKey}/`,
        (err, res) => {
            if (err){
                throw err;
            }
            else{
                var response = JSON.parse(res.body);
                scanResponseCode(
                    response.response_code,
                    (resErr) => {
                        callback(resErr, response);
                    }
                );
            }
        }    
    );
}

exports.trainFare = function(trainNo,srcStationCode,destStationCode,age,quotaCode,date,callback){
    request(
        `http://api.railwayapi.com/v2/fare/train/${trainNo}/source/${srcStationCode}/dest/${destStationCode}/age/${age}/quota/${quotaCode}/date/${date}/apikey/${apiKey}/`,
        (err, res) => {
            if (err){
                throw err;
            }
            else{
                var response = JSON.parse(res.body);
                scanResponseCode(
                    response.response_code,
                    (resErr) => {
                        callback(resErr, response);
                    }
                );
            }
        }    
    );
}

exports.trainArrivals = function(srcStationCode,windowHours,callback){
    request(
        `http://api.railwayapi.com/v2/arrivals/station/${srcStationCode}/hours/${windowHours}/apikey/${apiKey}/`,
        (err, res) => {
            if (err){
                throw err;
            }
            else{
                var response = JSON.parse(res.body);
                scanResponseCode(
                    response.response_code,
                    (resErr) => {
                        callback(resErr, response);
                    }
                );
            }
        }    
    );
}

exports.cancelledTrains = function(date,callback){
    request(
        `http://api.railwayapi.com/v2/cancelled/date/${date}/apikey/${apiKey}/`,
        (err, res) => {
            if (err){
                throw err;
            }
            else{
                var response = JSON.parse(res.body);
                scanResponseCode(
                    response.response_code,
                    (resErr) => {
                        callback(resErr, response);
                    }
                );
            }
        }    
    );
}

exports.rescheduledTrains = function(date,callback){
    request(
        `http://api.railwayapi.com/v2/rescheduled/date/${date}/apikey/${apiKey}/`,
        (err, res) => {
            if (err){
                throw err;
            }
            else{
                var response = JSON.parse(res.body);
                scanResponseCode(
                    response.response_code,
                    (resErr) => {
                        callback(resErr, response);
                    }
                );
            }
        }    
    );
}

exports.stationAutocomplete = function(partialStationName,callback){
    request(
        `http://api.railwayapi.com/v2/suggest-station/name/${partialStationName}/apikey/${apiKey}/`,
        (err, res) => {
            if (err){
                throw err;
            }
            else{
                var response = JSON.parse(res.body);
                scanResponseCode(
                    response.response_code,
                    (resErr) => {
                        callback(resErr, response);
                    }
                );
            }
        }    
    );
}

exports.trainAutocomplete = function(partialTrainNameOrNum,callback){
    request(
        `http://api.railwayapi.com/v2/suggest-train/train/${partialTrainNameOrNum}/apikey/${apiKey}/`,
        (err, res) => {
            if (err){
                throw err;
            }
            else{
                var response = JSON.parse(res.body);
                scanResponseCode(
                    response.response_code,
                    (resErr) => {
                        callback(resErr, response);
                    }
                );
            }
        }    
    );
}
